### PR TITLE
Add solver structure and tests

### PR DESCRIPTION
**Pull request by tmigot**
Make a `KnitroSolver` structure, which allows evaluating the model after a solve call.
This solves #55 .

The issue of solving the model again with a different solution remains as
```
ERROR: Both residuals and an objective function have been added to the model.
       The current version of Knitro does not support this.
```

Tested locally.

### CODE REVIEW

**Comment by dpo** [comment]
This should be split into 2 prs: one to restructure the files and one to introduce the solver.

**Comment by dpo** [comment]
Needs to be rebased.

**Comment by tmigot** [comment]
@JSOBot runtests

File: src/api.jl [hunk 1]
```diff
@@ -59,5 +59,95 @@ function knitro_statuses(code::Integer)
   return :unknown
 end
 
+mutable struct KnitroSolver
+  kc
+end
+
+function Base.finalize(solver::KnitroSolver)
+  KNITRO.KN_reset_params_to_defaults(solver.kc)
```
Comment (dpo) [inline src/api.jl:67]: Is this necessary?

File: src/api.jl [hunk 2]
```diff
@@ -59,5 +59,95 @@ function knitro_statuses(code::Integer)
...
+  callback::Union{Function, Nothing} = nothing,
+  kwargs...,
+)
+  solver = KnitroSolver(val, nlp, callback = callback; kwargs...)
+  stats = _knitro(nlp, solver)
+  finalize(solver)
+  return stats
+end
+
+function _knitro(nlp::AbstractNLPModel, solver::KnitroSolver)
```
Comment (dpo) [inline src/api.jl:120]: This should probably be `_knitro!` now?!

File: src/api.jl [hunk 3]
```diff
@@ -59,5 +59,95 @@ function knitro_statuses(code::Integer)
...
+
+  return solver
+end
+
+KnitroSolver(nlp::AbstractNLPModel; kwargs...) = KnitroSolver(_is_general_nlp(nlp), nlp; kwargs...)
+
 include("nlp.jl")
 include("nls.jl")
+
+knitro(nlp::AbstractNLPModel, solver::KnitroSolver) = _knitro(nlp, solver)
```
Comment (dpo) [inline src/api.jl:106]: This should probably be `knitro!` now?!

**Comment by tmigot** [comment]
@JSOBot runtests

File: src/api.jl (hunk 2 above)
Comment (tmigot) [reply src/api.jl]: you are right, thanks.

File: src/api.jl (hunk 1 above)
Comment (tmigot) [reply src/api.jl]: I removed it.

File: src/api.jl [hunk 4]
```diff
@@ -59,5 +59,96 @@ function knitro_statuses(code::Integer)
...
+end
+
+KnitroSolver(nlp::AbstractNLPModel; kwargs...) = KnitroSolver(_is_general_nlp(nlp), nlp; kwargs...)
+
 include("nlp.jl")
 include("nls.jl")
+
+knitro!(nlp::AbstractNLPModel, solver::KnitroSolver) = _knitro!(nlp, solver)
+
+function _knitro(
```
Comment (dpo) [inline src/api.jl:109]: Is this still used anywhere?

**Comment by tmigot** [comment]
@JSOBot runtests

File: src/api.jl (hunk 4 above)
Comment (tmigot) [reply src/api.jl]: To simplify, I removed `_knitro` and `_knitro!` as they don't seem to bring anything special.

**Comment by dpo** [comment]
Looks very natural. Thanks!

**Pull request merged**